cpp/include/cuspatial/experimental/detail/quadtree_bbox_filtering.cuh

```diff
@@ -0,0 +1,199 @@
+#include <cuspatial/experimental/detail/join/intersection.cuh>
+#include <cuspatial/experimental/detail/join/traversal.cuh>
+#include <cuspatial/experimental/point_quadtree.cuh>
+#include <cuspatial/traits.hpp>
+
+#include <thrust/iterator/discard_iterator.h>
+
+#include <iterator>
+#include <utility>
+
+namespace cuspatial {
+
+template <class KeyIterator,
+          class LevelIterator,
+          class IsInternalIterator,
+          class BoundingBoxIterator,
+          class T>
+std::pair<rmm::device_uvector<uint32_t>, rmm::device_uvector<uint32_t>>
+join_quadtree_and_bounding_boxes(KeyIterator keys_first,
+                                 KeyIterator keys_last,
+                                 LevelIterator levels_first,
+                                 IsInternalIterator is_internal_nodes_first,
+                                 KeyIterator lengths_first,
+                                 KeyIterator offsets_first,
+                                 BoundingBoxIterator bounding_boxes_first,
+                                 BoundingBoxIterator bounding_boxes_last,
+                                 T x_min,
+                                 T y_min,
+                                 T scale,
+                                 int8_t max_depth,
+                                 rmm::mr::device_memory_resource* mr,
+                                 rmm::cuda_stream_view stream)
+{
+  static_assert(is_same<T, cuspatial::iterator_vec_base_type<BoundingBoxIterator>>(),
+                "Iterator value_type mismatch");
+
+  auto const num_nodes = std::distance(keys_first, keys_last);
+  auto const num_boxes = std::distance(bounding_boxes_first, bounding_boxes_last);
+
+  // Count the number of top-level nodes to start.
+  // This could be provided explicitly, but count_if should be fast enough.
+  auto num_top_level_leaves = thrust::count_if(rmm::exec_policy(stream),
+                                               levels_first,
+                                               levels_first + num_nodes,
+                                               thrust::placeholders::_1 == 0);
+
+  auto num_pairs = num_top_level_leaves * num_boxes;
+
+  int32_t num_leaves{0};
+  int32_t num_results{0};
+  int32_t num_parents{0};
+
+  // The found bbox-quad pairs are dynamic and can not be pre-allocated.
+  // Relevant arrays are resized accordingly for memory efficiency.
+
+  // Vectors for intermediate bbox and node indices at each level
+  rmm::device_uvector<uint8_t> cur_types(num_pairs, stream);
+  rmm::device_uvector<uint8_t> cur_levels(num_pairs, stream);
+  rmm::device_uvector<uint32_t> cur_node_idxs(num_pairs, stream);
+  rmm::device_uvector<uint32_t> cur_bbox_idxs(num_pairs, stream);
+
+  // Vectors for found pairs of bbox and leaf node indices
+  rmm::device_uvector<uint32_t> out_node_idxs(num_pairs, stream, mr);
+  rmm::device_uvector<uint32_t> out_bbox_idxs(num_pairs, stream, mr);
+
+  auto make_current_level_iter = [&]() {
```

Comment:
Any reason to use a lambda here? It seems we should create the iterator as a variable instead.

Reply:
Yes, this lambda closes over variables that are updated each iteration of the loop.

Reply:
Do you mean that the iterators gets invalidated after every iteration of the loop?

Reply:
`make_current_level_iter()` returns an zip iterator over the `cur_{types,levels,node_idxs,bbox_idxs}` vectors, which represent the current level in the tree descent. The `descend_quadtree()` call [here](https://github.com/rapidsai/cuspatial/blob/adb747a4b4bb2d677b3d908cf3175879816bb27e/cpp/include/cuspatial/experimental/detail/quadtree_bbox_filtering.cuh#L143-L151) descends one level in the tree and returns vectors representing the next "current level." 

Similarly, `make_output_values_iter()` returns a zip iterator of `num_results + out_{node,bbox}_idxs`. `num_results` is updated by the `detail::find_intersections()` call [here](https://github.com/rapidsai/cuspatial/blob/adb747a4b4bb2d677b3d908cf3175879816bb27e/cpp/include/cuspatial/experimental/detail/quadtree_bbox_filtering.cuh#L154-L172).

The iterators returned by these functions are used both in setting up the initial state of the the top-level quadrants, and in the loop that does the tree traversal. An earlier implementation tracked and updated these two iterators each iteration of the loop, but [they were refactored](https://github.com/rapidsai/cuspatial/commit/d618baf989cd455ae65d90ddf102f455ca4eb90c#diff-c3817f6937f1738ced51e1fa383142ce02976d629b91fe5f19f4203eb63dc5b7) into these lambdas in a cleanup PR.